src/front/wgsl/mod.rs

```diff
@@ -2035,14 +2035,60 @@ impl Parser {
                 }
             }
         } else {
+            components.push(last_component);
+            let mut compose_components = Vec::new();
+
+            if let (
+                &crate::TypeInner::Matrix {
+                    rows,
+                    width,
+                    columns,
+                },
+                &crate::TypeInner::Scalar {
+                    kind: crate::ScalarKind::Float,
+                    ..
+                },
+            ) = (
+                ty_resolution.inner_with(ctx.types),
+                ctx.typifier.get(last_component, ctx.types),
+            ) {
+                let vec_ty = ctx.types.insert(
+                    crate::Type {
+                        name: None,
+                        inner: crate::TypeInner::Vector {
+                            width,
+                            kind: crate::ScalarKind::Float,
```

Comment:
We could just pass `kind` from the `TypeInner::Scalar` to here. Validation will take care of the rest.